File: sample-domain/npm/src/abstractions/unit-of-work.ts
import { IAsyncDisposable } from "./disposable";

export interface IUnitOfWork extends IAsyncDisposable {
    beginAsync(): Promise<string>;
    commitAsync(): Promise<void>;
    rollbackAsync(): Promise<void>;
}

export interface IUnitOfWorkFactory<TUnitOfWork extends IUnitOfWork> {
    create(): TUnitOfWork;
}

File: sample-domain/dotnet/src/abstractions/unit-of-work.cs
using System;
using System.Threading.Tasks;

namespace Sample.Domain.Abstractions
{
    public interface IUnitOfWork : IDisposable 
    { 
        Task<string> BeginAsync();
        Task CommitAsync();
        Task RollbackAsync();
    }

    public interface IUnitOfWorkFactory<TUnitOfWork> where TUnitOfWork : IUnitOfWork
    {
        TUnitOfWork Create();
    }
}